File: GA_01/GeneticAlgorithem.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.Random;

public class GeneticAlgorithem implements Comparator<DNA> {
	
	boolean endAlgorithm=false;
	ArrayList<DNA> population ;
	int Generation;
	int popSize;
	double mutationRate;
	Random random;
	ArrayList<DNA> newpopulation;
	char[] target;
	double fitnessSum;
	int dnaSize;
	DNA bestFitted;
	
	public GeneticAlgorithem(int popSize,int dnaSize,Random random,double mutationRate,char[] target){
		this.popSize= popSize;
		this.random=random;
		this.mutationRate=mutationRate;
		population = new ArrayList<DNA>();	
		Generation=1;
		newpopulation=new ArrayList<DNA>();
		this.dnaSize=dnaSize;
		//populate ...
		for(int i=0;i<popSize;i++){
			population.add(new DNA(dnaSize,target,random,true));
		}		
		bestFitted=new DNA(dnaSize,target,random,true);
	}
	public DNA weightedRandomSelection(){
		/*returns a DNA ,DNAes with higher fitness are more 
		 * Luckily to be chosen */
		fitnessSum=0;
		for (int j=0;j<popSize;j++){
			fitnessSum +=population.get(j).fitness;//get the total fitness 
		}
		for(int i=0;i<popSize;i++){
			double test=random.nextDouble()*fitnessSum;
			if(test<population.get(i).fitness)//if this DNA has a fitness larger than a random number*total fitness select it
				return population.get(i);
			test=test-population.get(i).fitness;//remove the selected DNA from the next calculation
		}		
		return population.get(random.nextInt(popSize)) ; 
}
	public void generateNewGenration(int elitism){	
			newpopulation.clear();
			//To Do ....ElEtism
			Collections.sort(population,this);
			Generation++;
		for(int i=0;i<popSize;i++){
			
			if(i<elitism){
				newpopulation.add(population.get(i));
			}
			else{
			DNA parent1= weightedRandomSelection();
			DNA parent2 = weightedRandomSelection();
			DNA child = parent1.crossOver(parent2);
			child.Mutate(mutationRate);
			child.calculateFitness();
			newpopulation.add(child);	
			}	
		}			
		population.clear();
		population.addAll(newpopulation);
		
		bestFitted.fitness=0;
		for(int i=0;i<popSize;i++){
			if(bestFitted.fitness < population.get(i).fitness)
				bestFitted=population.get(i);
			if(population.get(i).fitness==1)
				endAlgorithm=true;
			}		
		System.out.print("best of his generation :   ");
		System.out.print( bestFitted.genes);
		System.out.print(" fitness :" + bestFitted.fitness);
		System.out.println(" generation :"+Generation);	
	}
	@Override
	public int compare(DNA o1, DNA o2) {
		if(o1.fitness<o2.fitness){
			
			return 1;
		}else if(o1.fitness>o2.fitness){
			return -1;
		}

		return 0;
	}
	

}
	
	

File: GA_02/GeneticAlgorithem.cs
using System;
using System.Collections.Generic;
using UnityEngine;
public  class GeneticAlgorithem : MonoBehaviour, IComparer<Agent>
{
    public GameObject agent;
    public bool endAlgorithm = false;
    internal List<Agent>population;
    public int Generation;
    public int popSize;
    public double mutationRate;
    internal System.Random random;
    internal List<Agent> newpopulation;
    internal double fitnessSum;
    public int dnaSize;
    internal Agent bestFitted;

    List<Agent> grave;
    public void Awake()
    {
        grave= new List<Agent>();
        population = new List<Agent>();
        Generation = 1;
        newpopulation = new List<Agent>();
        random = new System.Random();

       
    }

    private void Start()
    {
        //populate
        for (int i = 0; i < popSize; i++)
        {
            Agent inst = Instantiate(agent).GetComponent<Agent>();
            inst.generate_random_chromosome(random);
            population.Add(inst);
        }
        //bestFitted = new Agent();
        StartRace();
        Time.timeScale = 3.5f;
    }


    public  Agent weightedRandomSelection()// choose parent 
    {
        /*returns a DNA ,DNAes with higher fitness are more 
		 * Luckily to be chosen */
        fitnessSum = 0;
        for (int j = 0; j < popSize; j++)
        {
            fitnessSum += population[j].fitness; //get the total fitness
        }
        for (int i = 0; i < popSize; i++)
        {
            //spin the fortune wheel 
            double spin = random.NextDouble() * fitnessSum;
            if (spin < population[i].fitness) //the larger the fitness of the indivedual the more luckely it's choosen 
            {
                return population[i];
            }
            spin = spin - population[i].fitness; //remove the selected DNA from the next calculation
        }
        return population[random.Next(popSize)];
    }
    public virtual void generateNewGenration(int elitism)//elitism : number of agents not effected 
    {
        newpopulation.Clear();
        population.Sort(this);//sort by fitness 
        Generation++;
        for (int i = 0; i < popSize; i++)
        {
            if (i <elitism)
            {
                print("&^&^&^&^&^"+population.Count);
                Agent topAgent = population[i];
                var clone = Instantiate(agent).GetComponent<Agent>();
                clone.SetChromosome(topAgent.chromosome);
                newpopulation.Add(clone);
                
            }
            else
            {
                Agent parent1 = weightedRandomSelection();
                Agent parent2 = weightedRandomSelection();
                Agent child =Instantiate(agent).GetComponent<Agent>();
                child.SetChromosome(parent1.crossOver(parent2.chromosome,random));
                child.Mutate(mutationRate,random);
                newpopulation.Add(child);
                
            }
        }

        for (int i = 0; i < popSize; i++)//delete gameobjects from scene
        {
            Destroy(population[i].gameObject);
        }
        population.Clear();

        foreach (var item in newpopulation)
        {
            population.Add(item);
        }
        bestFitted = population[0];
        print("best of his generation :   ");
        print("best chromosome"+bestFitted.chromosome.ToArray());
        print(" fitness :" + bestFitted.fitness);
        print(" generation :" + Generation);

    }

void StartRace()
    {
        for (int i = 0; i <popSize; i++)
        {
            population[i]._start = true;
        }

    }
    public virtual int Compare(Agent o1,Agent o2)
    {
        if (o1.fitness < o2.fitness)
        {
            return 1;
        }
        else if (o1.fitness > o2.fitness)
        {
            return -1;
        }
        return 0;
    }
    public void agent_died(Agent _agent) {
        grave.Add(_agent);
        print("OH Iam DEAD , MY FITNESS IS :"+_agent.fitness+"  dead counter:"+ grave.Count);
       
        if (grave.Count==popSize)
        {
            grave.Clear();
            generateNewGenration(2); 
            StartRace();
        }
    }
}
